### PR TITLE
[Feature] New RBAC privilege framework: create & drop role

### PR DESCRIPTION
**Pull request by padmejin**
## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

Introduce a new privilege framework.
To avoid huge PR which is hard to review and easy to conflict, I'm splitting into several small PR, each implementing an individual aspect of the new privilege framework. This is the seventh one. For previous PRs see 
https://github.com/StarRocks/starrocks/pull/11447
https://github.com/StarRocks/starrocks/pull/11528
https://github.com/StarRocks/starrocks/pull/11593
https://github.com/StarRocks/starrocks/pull/11732
https://github.com/StarRocks/starrocks/pull/11777
https://github.com/StarRocks/starrocks/pull/11847

This PR adds a map in PrivilegeManager to maintain roles.
* support create & drop role
* persist roles to image & journal.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 95 / 124 (76.61%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|0|3|00.00%|[733, 734, 735]|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|5|13|38.46%|[926, 927, 928, 929, 932, 933, 934, 935]|
|:large_blue_circle:|com/starrocks/persist/RolePrivilegeCollectionInfo.java|8|14|57.14%|[38, 42, 46, 50, 59, 60]|
|:large_blue_circle:|com/starrocks/qe/DDLStmtExecutor.java|4|6|66.67%|[449, 461]|
|:large_blue_circle:|com/starrocks/privilege/PrivilegeManager.java|42|51|82.35%|[344, 346, 347, 349, 351, 369, 371, 373, 375]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeStmtAnalyzerV2.java|15|16|93.75%|[105]|
|:large_blue_circle:|com/starrocks/privilege/RolePrivilegeCollection.java|21|21|100.00%|[]|

**Comment by padmejin** [comment]
run all

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 144 / 155 (92.90%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|5|13|38.46%|[926, 927, 928, 929, 932, 933, 934, 935]|
|:large_blue_circle:|com/starrocks/qe/DDLStmtExecutor.java|6|8|75.00%|[450, 462]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeStmtAnalyzerV2.java|15|16|93.75%|[105]|
|:large_blue_circle:|com/starrocks/persist/RolePrivilegeCollectionInfo.java|14|14|100.00%|[]|
|:large_blue_circle:|com/starrocks/privilege/PrivilegeManager.java|77|77|100.00%|[]|
|:large_blue_circle:|com/starrocks/privilege/RolePrivilegeCollection.java|24|24|100.00%|[]|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|3|3|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/persist/EditLog.java [hunk 1]
```diff
@@ -1577,4 +1590,18 @@ public void logUpdateUserPrivilege(
...
+            long roleId,
+            RolePrivilegeCollection privilegeCollection,
+            short pluginId,
+            short pluginVersion) {
+        RolePrivilegeCollectionInfo info = new RolePrivilegeCollectionInfo(
+                roleId, privilegeCollection, pluginId, pluginVersion);
+        logEdit(OperationType.OP_UPDATE_ROLE_PRIVILEGE_V2, info);
+    }
+
+    public void logDropRole(long roleId) {
```
Comment (gengjun-git) [inline fe/fe-core/src/main/java/com/starrocks/persist/EditLog.java:1604]: Best to use json KV format, even though there is only one property. To support expand properties.

File: fe/fe-core/src/main/java/com/starrocks/privilege/PrivilegeManager.java [hunk 2]
```diff
@@ -48,9 +50,16 @@ protected PrivilegeManager() {
...
         userToPrivilegeCollection = new HashMap<>();
+        roleIdToPrivilegeCollection = new HashMap<>();
         userLock = new ReentrantReadWriteLock();
+        roleLock = new ReentrantReadWriteLock();
+        nextRoleId = 0;
     }
 
+    private Map<Long, RolePrivilegeCollection> roleIdToPrivilegeCollection;
+    @SerializedName(value = "r")
+    private long nextRoleId;
```
Comment (gengjun-git) [inline fe/fe-core/src/main/java/com/starrocks/privilege/PrivilegeManager.java:61]: Why not using GlobalStateMgr.getNextId()?

File: fe/fe-core/src/main/java/com/starrocks/privilege/PrivilegeManager.java [hunk 3]
```diff
@@ -380,18 +500,28 @@ public static PrivilegeManager load(
...
+                // 1 json for num roles
+                int numRole = (int) reader.readJson(int.class);
+                LOG.info("loading {} roles", numRole);
+                for (int i = 0; i != numRole; ++i) {
+                    // 2 json for each role(kv)
+                    Long roleId = (Long) reader.readJson(Long.class);
+                    RolePrivilegeCollection collection =
+                            (RolePrivilegeCollection) reader.readJson(RolePrivilegeCollection.class);
+                    ret.roleIdToPrivilegeCollection.put(roleId, collection);
+                }
```
Comment (gengjun-git) [inline fe/fe-core/src/main/java/com/starrocks/privilege/PrivilegeManager.java:512]: You should update the nextRoleId here.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 143 / 154 (92.86%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|6|14|42.86%|[926, 927, 928, 929, 932, 933, 934, 935]|
|:large_blue_circle:|com/starrocks/qe/DDLStmtExecutor.java|4|6|66.67%|[450, 462]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeStmtAnalyzerV2.java|15|16|93.75%|[106]|
|:large_blue_circle:|com/starrocks/persist/RolePrivilegeCollectionInfo.java|14|14|100.00%|[]|
|:large_blue_circle:|com/starrocks/privilege/PrivilegeManager.java|77|77|100.00%|[]|
|:large_blue_circle:|com/starrocks/privilege/RolePrivilegeCollection.java|24|24|100.00%|[]|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|3|3|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/privilege/PrivilegeManager.java (hunk 2 above)
Comment (padmejin) [reply fe/fe-core/src/main/java/com/starrocks/privilege/PrivilegeManager.java]: done

File: fe/fe-core/src/main/java/com/starrocks/privilege/PrivilegeManager.java (hunk 3 above)
Comment (padmejin) [reply fe/fe-core/src/main/java/com/starrocks/privilege/PrivilegeManager.java:547]: nextRoleId used to store in the first part, but I have removed it anyway.

**Review by gengjun-git — APPROVED** [review]

**Review by waittttting — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

File: fe/fe-core/src/main/java/com/starrocks/privilege/PrivilegeManager.java [hunk 4]
```diff
@@ -306,6 +331,93 @@ public short analyzeType(String typeName) throws PrivilegeException {
...
 
+    public void createRole(CreateRoleStmt stmt) throws DdlException {
+        roleWriteLock();
+        try {
+            String roleName = stmt.getQualifiedRole();
+            Long roleId = getRoleIdByNameNoLock(roleName);
+            if (roleId != null) {
+                throw new DdlException(String.format("Role %s already exists! id = %d", roleName, roleId));
+            }
+            RolePrivilegeCollection collection = new RolePrivilegeCollection(roleName);
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/privilege/PrivilegeManager.java:342]: so we don't support create role with some initial priviledges? we can only grant roles or priviledges to a role after creation?

Comment (padmejin) [reply fe/fe-core/src/main/java/com/starrocks/privilege/PrivilegeManager.java:342]: No, we don't. Neither will we support creating a user with initial privileges. PM had a discussion earlier and decided it's better this way.

Comment (nshangyiming) [reply fe/fe-core/src/main/java/com/starrocks/privilege/PrivilegeManager.java:342]: ok

**Review by nshangyiming — APPROVED** [review]

**Comment by padmejin** [comment]
run starrocks_admit_test

**Review by meegoo — APPROVED** [review]

**Pull request merged**